TipCatDotNet.Api/Services/Auth/InvitationService.cs

```diff
@@ -105,33 +87,84 @@ public async Task<Result> Redeem(int memberId, CancellationToken cancellationTok
+        public Task<Result> Send(MemberContext memberContext, MemberRequest request, CancellationToken cancellationToken = default)
+        {
+            return Validate()
+                .Bind(EnsureInvitationExists)
+                .Bind(invitation => SendInternal(request, invitation, cancellationToken));
+
+
+            Result Validate()
+            {
+                var validator = new MemberRequestValidator(memberContext, _context);
+                return validator.ValidateInvite(request).ToResult();
+            }
+
+
+            async Task<Result<MemberInvitation>> EnsureInvitationExists()
+            {
+                var existedInvitation = await _context.MemberInvitations
+                    .Where(i => i.MemberId == request.Id!)
+                    .SingleOrDefaultAsync(cancellationToken);
+
+                if (existedInvitation is null)
+                    return Result.Failure<MemberInvitation>($"No invitations found for a member {request.Id!}, or they expired.");
+
+                if (existedInvitation.State == InvitationStates.Accepted)
+                    return Result.Failure<MemberInvitation>("The invitation was accepted already.");
+
+                return existedInvitation;
+            }
+        }
+
+        
+        private Task<Result> SendInternal(MemberRequest request, MemberInvitation memberInvitation, CancellationToken cancellationToken)
         {
-            var link = await _context.MemberInvitations
-                .Where(i => i.MemberId == memberId 
-                    && (i.State == InvitationStates.NotSent || i.State == InvitationStates.Sent)
-                    && DateTime.UtcNow.AddDays(-1) <= i.Created)
-                .Select(i => i.Link)
-                .SingleOrDefaultAsync(cancellationToken);
-
-            if (link is null)
-                return Result.Failure($"No invitations found for a member {memberId}, or they expired.");
-
-            var member = await _context.Members
-                .Where(m => m.Id == memberId)
-                .Select(m => new { m.Email, m.AccountId })
-                .SingleOrDefaultAsync(cancellationToken);
-
-            if (member is null)
-                return Result.Failure($"No email addresses found for a member {memberId}.");
-
-            var accountName = await _context.Accounts
-                .Where(a => a.Id == member.AccountId)
-                .Select(a => a.OperatingName)
-                .SingleOrDefaultAsync(cancellationToken);
-
-            return await _mailSender.Send(_options.TemplateId, member.Email!,
-                new MemberInvitationEmail(accountName, link, CompanyInfoService.Get));
+            var now = DateTime.UtcNow;
+
+            return ChangePasswordIfNeeded()
+                .Bind(AddLinkAndMarkAsSent)
+                .Bind(ResendEmail);
+
+
+            async Task<Result<MemberInvitation>> ChangePasswordIfNeeded()
+            {
+                if (now.AddDays(-7) >= memberInvitation.Created)
```

Comment:
If memberInvitation was created more than seven days ago, change password not needed and we return existing memberInvitation. Otherwise we change password?

Reply:
Auth0 ticket time frame is 7 days, so I just clear unused invitations that way. I'll ad a comment

Reply:
Thank you)